### ISSUE TITLE
`npm run webpack` failing on Linux

### ISSUE THREAD

**Issue by uwagura** [bartfeenstra/betty#2177]
Hello Bart. 

I recently tried running betty on a machine running Fedora. When I call `betty -c betty.yaml generate` on this machine, I get the following error (truncated for readability)
```
    | Traceback (most recent call last):
    |   File "<path_to_betty_install>/lib64/python3.12/site-packages/betty/extension/webpack/__init__.py", line 172, in generate
    |     build_directory_path = await self._generate_ensure_build_directory(
    |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "<path_to_betty_install>/lib64/python3.12/site-packages/betty/extension/webpack/__init__.py", line 233, in _generate_ensure_build_directory
    |     return await builder.build()
    |            ^^^^^^^^^^^^^^^^^^^^^
    |   File "<path_to_betty_install>/lib64/python3.12/site-packages/betty/extension/webpack/build.py", line 262, in build
    |     await self._webpack_build(
    |   File "<path_to_betty_install>/lib64/python3.12/site-packages/betty/extension/webpack/build.py", line 234, in _webpack_build
    |     await _npm.npm(("run", "webpack"), cwd=npm_project_directory_path)
    |   File "<path_to_betty_install>/lib64/python3.12/site-packages/betty/_npm.py", line 42, in npm
    |     return await run_process(
    |            ^^^^^^^^^^^^^^^^^^
    |   File "<path_to_betty_install>lib64/python3.12/site-packages/betty/subprocess.py", line 54, in run_process
    |     raise CalledProcessError(
    | subprocess.CalledProcessError: Command 'npm run webpack' returned non-zero exit status 127.

```
I installed npm using the instructions for linux [here](https://nodejs.org/en/download/package-manager), and confirmed that I can access npm via `npm --help`. When I run `betty -vvv -c betty.yaml generate`, I get the following output: 
```
Running subprocess `npm run webpack`...
Subprocess `npm run webpack` stdout:
> webpack
> webpack --config webpack.config.js
Subprocess `npm run webpack` stderr:
sh: line 1: webpack: command not found

``` 
And interestingly, I get the same exact output even if I run `npm install webpack`. This is the same yaml that successfully generated a site on my windows machine, so I'm not sure what's causing this issue. Is there anything else I need to do to use Betty on linux? Here is the output from `betty --version`, let me know  if there is something I'm missing. Thanks!
```
Betty                   0.3.14
Operating system        Linux-6.8.5-301.fc40.x86_64-x86_64-with-glibc2.39
Python                  3.12.7 (main, Oct  1 2024, 00:00:00) [GCC 14.2.1 20240912 (Red Hat 14.2.1-3)]
```

**Comment by bartfeenstra** [comment]
HI @uwagura ! Thanks for reporting this. There's a lot of good detail in your post, but I'm not immediately sure what may cause this.

First of all, any npm dependencies are managed by Betty (in fact, the Webpack extension that's used internally). You will not have to call npm or install anything yourself. In fact, that's rather hard, because the necessary `package.json` files are deep within the Betty codebase. Running npm within your project directory will not have any effect as that's not how Betty handles this internally (unlike if Betty had been an actual Node.js project).

Can you provide the following?
- The full `betty --version` output
- The full output of running `betty generate` with the `-vv` option, which will show debugging messages that may shed light on the problem

**Comment by uwagura** [comment]
Full output from `betty -vvv -c betty.yaml generate`: 
```
Using selector: EpollSelector
Using selector: EpollSelector
Using selector: EpollSelector
Running subprocess `npm --version`...
`npm` is available
Using selector: EpollSelector
Loaded the configuration from <path_to_betty>/betty_gui.yaml.
Using selector: EpollSelector
Using selector: EpollSelector
Loading "<path_to_betty>/tree.gramps"...
Loaded 22 notes.
Loaded 22 files.
Loaded 5 repositories as sources.
Loaded 3 sources.
Loaded 17 citations.
Loaded 6 places.
Loaded 18 events.
Loaded 47 people.
Generating your site to <path_to_betty>/output.
Generating static public files...
Using selector: EpollSelector
Generating JSON Schema...
Generating OpenAPI specification...
Generating localized public files in English (United States)...
Generated 0 out of 293 items (0%).
Running subprocess `npm run webpack`...
Subprocess `npm run webpack` stdout:

> webpack
> webpack --config webpack.config.js

Subprocess `npm run webpack` stderr:
sh: line 1: webpack: command not found

unhandled errors in a TaskGroup (1 sub-exception)
  + Exception Group Traceback (most recent call last):
  |   File "<path_to_betty>/bettyenv/lib64/python3.12/site-packages/betty/cli.py", line 64, in catch_exceptions
  |     yield
  |   File "/usr/lib64/python3.12/contextlib.py", line 81, in inner
  |     return func(*args, **kwds)
  |            ^^^^^^^^^^^^^^^^^^^
  |   File "<path_to_betty>/bettyenv/lib64/python3.12/site-packages/betty/cli.py", line 109, in _command
  |     return run(__command())
  |            ^^^^^^^^^^^^^^^^
  |   File "/usr/lib64/python3.12/asyncio/runners.py", line 194, in run
  |     return runner.run(main)
  |            ^^^^^^^^^^^^^^^^
  |   File "/usr/lib64/python3.12/asyncio/runners.py", line 118, in run
  |     return self._loop.run_until_complete(task)
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/usr/lib64/python3.12/asyncio/base_events.py", line 687, in run_until_complete
  |     return future.result()
  |            ^^^^^^^^^^^^^^^
  |   File "<path_to_betty>/bettyenv/lib64/python3.12/site-packages/betty/cli.py", line 107, in __command
  |     await f(app, *args, **kwargs)
  |   File "<path_to_betty>/bettyenv/lib64/python3.12/site-packages/betty/cli.py", line 327, in _generate
  |     await generate.generate(app)
  |   File "<path_to_betty>/bettyenv/lib64/python3.12/site-packages/betty/generate.py", line 119, in generate
  |     await completed_job
  |   File "/usr/lib64/python3.12/asyncio/tasks.py", line 631, in _wait_for_one
  |     return f.result()  # May raise f.exception().
  |            ^^^^^^^^^^
  |   File "<path_to_betty>/bettyenv/lib64/python3.12/site-packages/betty/generate.py", line 168, in _job
  |     await f(*args, **kwargs)
  |   File "<path_to_betty>/bettyenv/lib64/python3.12/site-packages/betty/generate.py", line 250, in _generate_dispatch
  |     await app.dispatcher.dispatch(Generator)(job_context)
  |   File "<path_to_betty>/bettyenv/lib64/python3.12/site-packages/betty/app/extension/__init__.py", line 443, in _dispatch
  |     for result in await gather(
  |                   ^^^^^^^^^^^^^
  |   File "<path_to_betty>/bettyenv/lib64/python3.12/site-packages/betty/asyncio.py", line 34, in gather
  |     async with TaskGroup() as task_group:
  |                ^^^^^^^^^^^
  |   File "/usr/lib64/python3.12/asyncio/taskgroups.py", line 145, in __aexit__
  |     raise me from None
  | ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "<path_to_betty>/bettyenv/lib64/python3.12/site-packages/betty/extension/webpack/__init__.py", line 172, in generate
    |     build_directory_path = await self._generate_ensure_build_directory(
    |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "<path_to_betty>/bettyenv/lib64/python3.12/site-packages/betty/extension/webpack/__init__.py", line 233, in _generate_ensure_build_directory
    |     return await builder.build()
    |            ^^^^^^^^^^^^^^^^^^^^^
    |   File "<path_to_betty>/bettyenv/lib64/python3.12/site-packages/betty/extension/webpack/build.py", line 262, in build
    |     await self._webpack_build(
    |   File "<path_to_betty>/bettyenv/lib64/python3.12/site-packages/betty/extension/webpack/build.py", line 234, in _webpack_build
    |     await _npm.npm(("run", "webpack"), cwd=npm_project_directory_path)
    |   File "<path_to_betty>/bettyenv/lib64/python3.12/site-packages/betty/_npm.py", line 42, in npm
    |     return await run_process(
    |            ^^^^^^^^^^^^^^^^^^
    |   File "<path_to_betty>/bettyenv/lib64/python3.12/site-packages/betty/subprocess.py", line 54, in run_process
    |     raise CalledProcessError(
    | subprocess.CalledProcessError: Command 'npm run webpack' returned non-zero exit status 127.
    +------------------------------------
Using selector: EpollSelector
```
Full output from `betty --version`
```
Betty                   0.3.14
Operating system        Linux-6.8.5-301.fc40.x86_64-x86_64-with-glibc2.39
Python                  3.12.7 (main, Oct  1 2024, 00:00:00) [GCC 14.2.1 20240912 (Red Hat 14.2.1-3)]
Python packages         aiofiles                             23.2.1
                        aiohappyeyeballs                     2.4.3
                        aiohttp                              3.10.10
                        aiosignal                            1.3.1
                        alabaster                            0.7.16
                        attrs                                24.2.0
                        babel                                2.16.0
                        beautifulsoup4                       4.12.3
                        betty                                0.3.14
                        certifi                              2024.8.30
                        charset-normalizer                   3.4.0
                        click                                8.1.7
                        docker                               7.1.0
                        docutils                             0.21.2
                        frozenlist                           1.5.0
                        furo                                 2024.5.6
                        geographiclib                        2.0
                        geopy                                2.4.1
                        html5lib                             1.1
                        idna                                 3.10
                        imagesize                            1.4.1
                        Jinja2                               3.1.4
                        jsonschema                           4.23.0
                        jsonschema-specifications            2024.10.1
                        langcodes                            3.4.1
                        language_data                        1.2.0
                        marisa-trie                          1.2.1
                        MarkupSafe                           2.1.5
                        multidict                            6.1.0
                        packaging                            24.1
                        pdf2image                            1.17.0
                        pillow                               10.4.0
                        pip                                  24.3.1
                        polib                                1.2.0
                        propcache                            0.2.0
                        Pygments                             2.18.0
                        PyQt6                                6.7.1
                        PyQt6-Qt6                            6.7.3
                        PyQt6_sip                            13.8.0
                        PyYAML                               6.0.2
                        referencing                          0.35.1
                        requests                             2.32.3
                        rpds-py                              0.20.0
                        setuptools                           75.2.0
                        six                                  1.16.0
                        snowballstemmer                      2.2.0
                        soupsieve                            2.6
                        Sphinx                               7.4.7
                        sphinx-autodoc-typehints             2.3.0
                        sphinx-basic-ng                      1.0.0b2
                        sphinx_design                        0.6.1
                        sphinxcontrib-applehelp              2.0.0
                        sphinxcontrib-devhelp                2.0.0
                        sphinxcontrib-htmlhelp               2.1.0
                        sphinxcontrib-jsmath                 1.0.1
                        sphinxcontrib-qthelp                 2.0.0
                        sphinxcontrib-serializinghtml        2.0.0
                        typing_extensions                    4.12.2
                        urllib3                              2.2.3
                        webencodings                         0.5.1
                        yarl                                 1.17.0

```

**Comment by bartfeenstra** [comment]
Thanks! I tried reproducing this locally on Ubuntu but so far it all runs fine.

While I have a think about this, can you run `betty clear-caches` and try again? It doesn't immediately look like it but it could be Betty is using an outdated or broken cache directory (in which it manages the npm packages internally).

**Comment by uwagura** [comment]
@bartfeenstra , `betty clear-caches` resolved this issue! I transferred a lot of my betty files from my windows machine to this Linux one, I guess a cache directory came along for the ride and was causing this issue. Thanks again for your help!

**Comment by bartfeenstra** [comment]
Phew! That made thing easy. Glad things are working for you now :)

0.4.0 beta 1 was released earlier, with 0.4.0 (stable) planned for the next 4-6 weeks. You may be interested in using that version if you just started a new project.
